# Project: Organizing Netflix Movie Actor Rating Data

## Analysis Objectives

Organize the average IMDB ratings of films in different genres, such as comedy, action, and science fiction, to identify actors in high-rated performances within each genre.  
Practice organizing data to obtain data suitable for further analysis.

## Introduction

The original dataset records all Netflix TV shows and films available in the United States up to July 2022. The dataset consists of two tables: `titles.csv` and `credits.csv`.  
`titles.csv` contains information related to films and TV shows, including IDs, titles, types (film or TV show), descriptions, genres, IMDB ratings, and more.  
`credits.csv` includes information about over 70,000 directors and actors who appear in Netflix productions, including names, title IDs, character names, roles (director/actor), and other relevant details.

Here are the meanings of each column in `titles.csv`:
- id: ID of the film or TV show.
- title: Title of the film or TV show.
- show_type: Type of the show, either "TV Show" or "Film".
- description: Brief description of the plot.
- release_year: Year of release.
- age_certification: Age certification rating.
- runtime: Duration of each episode for TV shows or the runtime for films.
- genres: List of genres.
- production_countries: List of countries where the production took place.
- seasons: Number of seasons (for TV shows).
- imdb_id: IMDB ID of the title.
- imdb_score: IMDB rating of the title.
- imdb_votes: Number of IMDB votes for the title.
- tmdb_popularity: Popularity score on TMDB.
- tmdb_score: TMDB score of the title.

And for `credits.csv`:
- person_ID: ID of the cast or crew member.
- id: ID of the film or TV show they participated in.
- name: Name of the person.
- character_name: Name of the character they portrayed.
- role: Role of the person, either "Actor" or "Director".

## Read the Data

Import the necessary libraries for data analysis and use Pandas' `read_csv` function to parse the contents of the raw data file "titles.csv" into a DataFrame assigned to the variable `original_titles`. Parse the contents of the raw data file "credits.csv" into a DataFrame assigned to the variable `original_credits`.

In [68]:
import pandas as pd

In [69]:
original_titles = pd.read_csv('titles.csv')
original_credits = pd.read_csv('credits.csv')

In [70]:
original_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [71]:
original_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

## Data Evaluation and Cleaning

The step will focus on two main aspects: structure and content, addressing tidiness and cleanliness.  
Structural issues pertain to data that does not adhere to the principles of "each variable as a column, each observation as a row, and each type of observational unit forms a table." Content issues include missing data, duplicate data, invalid data, and other data quality concerns.

To distinguish between cleaned and original data, we will create new variables `cleaned_titles`, which will be a copy of `original_titles`, and `cleaned_credits`, which will be a copy of `original_credits`. Subsequent cleaning steps will be applied to `cleaned_titles` and `cleaned_credits`.

In [72]:
cleaned_titles = original_titles.copy()
cleaned_credits = original_credits.copy()

### Data Tidiness

In [73]:
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   
5   ts22164         Monty Python's Flying Circus   SHOW   
6   tm70993                        Life of Brian  MOVIE   
7   tm14873                          Dirty Harry  MOVIE   
8  tm119281                     Bonnie and Clyde  MOVIE   
9   tm98978                      The Blue Lagoon  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   
5  A British sketch comedy series with the shows ...          1969   
6  Brian Cohen is an average young Jewish man, bu...          1979   
7  When a madman dubbed 'Scorpio' terrorizes San ...          1971   
8  In the 1930s, bored waitress Bonnie Parker fal...          1967   
9  Two small children and a ship's cook survive a...          1980   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   
5             TV-14       30                       ['comedy', 'european']   
6                 R       94                                   ['comedy']   
7                 R      102              ['thriller', 'action', 'crime']   
8                 R      110                 ['crime', 'drama', 'action']   
9                 R      104               ['romance', 'action', 'drama']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   
5               ['GB']      4.0  tt0063929         8.8     73424.0   
6               ['GB']      NaN  tt0079470         8.0    395024.0   
7               ['US']      NaN  tt0066999         7.7    155051.0   
8               ['US']      NaN  tt0061418         7.7    112048.0   
9               ['US']      NaN  tt0080453         5.8     69844.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600  
5           17.617       8.306  
6           17.770       7.800  
7           12.817       7.500  
8           15.687       7.500  
9           50.324       6.156

From examining the first 10 rows of the data, it appears that the `genres` and `production_countries` variables in `cleaned_titles` contain multiple values that need to be split.
Let's first extract one of the values from the `genres` variable for observation.

In [74]:
cleaned_titles['genres'][0]

"['documentation']"

Although `genres` appears to be in list format, its actual type is string, not a list of strings, which prevents us from directly using `value_counts` to count occurrences of each value. We can utilize Python's built-in `eval` function, which can convert strings to expressions, to convert the string representation of a list into an actual list.

In [75]:
cleaned_titles['genres'] = cleaned_titles['genres'].apply(eval)
cleaned_titles['genres'][0]

['documentation']

Once converted into a list, we can use the `explode` method of the DataFrame to split the column's list values into separate rows.

In [76]:
cleaned_titles = cleaned_titles.explode('genres')
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   
2                 R      109       thriller               ['US']      NaN   
2                 R      109       european               ['US']      NaN   
3                PG       91        fantasy               ['GB']      NaN   
3                PG       91         action               ['GB']      NaN   
3                PG       91         comedy               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

Next, the process will be the same for the `production_countries` column.  
Each observation's `production_countries` value does not represent a single country but rather a list of countries. Let's first extract a value from the `production_countries` variable for observation.

In [77]:
cleaned_titles['production_countries'][0]

"['US']"

As we can see, the `production_countries` column faces a similar issue. Although it appears in list format, its actual type is not a list of strings but a string, making it difficult to split. We can again utilize the `eval` function for type conversion and verify that it indeed becomes a list type after conversion.

In [78]:
cleaned_titles['production_countries'] = cleaned_titles['production_countries'].apply(lambda x : eval(x))
cleaned_titles['production_countries'][0]

['US']

Once the type conversion is confirmed, we will use the `explode` method to split the list values into separate rows.

In [79]:
cleaned_titles = cleaned_titles.explode('production_countries')
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   
2                 R      109       thriller                   US      NaN   
2                 R      109       european                   US      NaN   
3                PG       91        fantasy                   GB      NaN   
3                PG       91         action                   GB      NaN   
3                PG       91         comedy                   GB      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

After addressing the structural issues in `cleaned_titles`, let's examine `cleaned_credits`.

In [80]:
cleaned_credits.sample(10)

person_id        id               name  character      role
2543       10782  tm116488     David Pasquesi  Cameraman     ACTOR
77542    1913908  ts287729  Gokul J. Krishnan       Self     ACTOR
11223     297627  tm101684             Ketche        NaN  DIRECTOR
42359     246698  tm332500      Lemuel Pelayo     Harold     ACTOR
25150      66871  tm208713      Shruti Haasan     Shruti     ACTOR
75230     289879  ts285854       Carol Abboud        NaN     ACTOR
31383       1114  tm368241        Mae Whitman      Ellen     ACTOR
69737     228434  tm996676        Sai Pallavi    Vennela     ACTOR
1896        3197  tm122434          Kirk Ward       Earl     ACTOR
26653      69505  tm201855      Vivan Bhatena      Ricky     ACTOR

Based on a sample of 10 rows, the data in `cleaned_credits` adheres to the principle of "each variable as a column, each observation as a row, and each type of observational unit forms a table," indicating no structural issues.

### Data Cleanliness

Use the `info` method to get a general understanding of the data content.

In [81]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

From the output, it appears that the `cleaned_titles` data contains 17,818 observations. Variables such as `title`, `description`, `age_certification`, `genres`, `production_countries`, `seasons`, `imdb_id`, `imdb_score`, `imdb_votes`, `tmdb_popularity`, and `tmdb_score` have missing values, which will be assessed and cleaned in subsequent steps.
Additionally, `release_year` represents a year and should not be numeric; it needs to be converted to a date format.

In [82]:
cleaned_titles["release_year"] = pd.to_datetime(cleaned_titles["release_year"], format='%Y')
cleaned_titles["release_year"]

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

In [83]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


From the output, it appears that `cleaned_credits` data contains 77,801 observations. The `character` variable has missing values, which will be assessed and cleaned in subsequent steps.  
Additionally, `person_id` represents the ID of cast or crew members and should not be numeric; it needs to be converted to a string data type.

In [84]:
cleaned_credits['person_id'] =  cleaned_credits['person_id'].astype('str')
cleaned_credits["person_id"].dtype

dtype('O')

#### Handling Missing Data

In `cleaned_titles`, there are missing values in the `title`, `description`, `age_certification`, `genres`, `production_countries`, `seasons`, `imdb_id`, `imdb_score`, `imdb_votes`, `tmdb_popularity`, and `tmdb_score`  variables.
Since the title, description, age certification, production countries, seasons, IMDB ID, IMDB votes, TMDB popularity, and TMDB score variables do not affect our analysis of identifying actors in high IMDB-rated films within each genre, we can retain observations with missing values for these variables.
However, `imdb_score` and `genres`, which represent IMDB ratings and genres, are crucial for our subsequent analysis.
Let's first extract observations with missing `imdb_score` values for examination.

In [85]:
cleaned_titles.query('imdb_score.isnull()')

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

Due to the missing core data `imdb_score` required for analysis, we will delete these observations and then examine the number and sum of missing values in that column after deletion.

In [86]:
cleaned_titles.dropna(subset=['imdb_score'], inplace=True)
cleaned_titles['imdb_score'].isnull().sum()

0

Next, let's extract observations with missing values in the `genres` column for examination.

In [87]:
cleaned_titles[cleaned_titles.genres.isnull()]

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description release_year  \
1813  TV legend David Letterman teams up with fascin...   2018-01-01   
1939  MInecraft: Story Mode is an interactive, anima...   2018-01-01   
2386  In this unscripted series starring comedy lege...   2018-01-01   
2658  #ABtalks is a YouTube interview show hosted by...   2018-01-01   
4274                                                NaN   2020-01-01   
4648  'In Vitro' is an otherworldly rumination on me...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

Due to the missing core data `genres` required for analysis, we will delete these observations and then examine the number and sum of missing values in that column after deletion.

In [88]:
cleaned_titles = cleaned_titles.dropna(subset=['imdb_score'])
cleaned_titles['imdb_score'].isnull().sum()

0

Next, let's assess the missing data in `cleaned_credits`, where only the `character` variable has missing values.
The character names do not affect our analysis of identifying actors in high IMDB-rated films within each genre. Additionally, missing values in this variable could be due to the individuals being directors or other roles where character names are not applicable. Therefore, we can retain observations with missing values in the `character` variable.

#### Handling Duplicate Data

Based on the meanings and contents of the data variables in `cleaned_titles`, there should not be any observations where all variable values are identical. Let's check for the presence of duplicate values.

In [89]:
cleaned_titles.duplicated().sum()

0

The output result of 0 indicates that there are no duplicate values in `cleaned_titles`.
Next, let's check if there are any duplicate values in the `cleaned_credits` dataset.

In [90]:
cleaned_credits.duplicated().sum()

0

The output result of 0 indicates that there are no duplicate values in the `cleaned_credits` dataset as well.

#### Handling Inconsistent Data

For `cleaned_titles`, inconsistencies may exist in the `genres` and `production_countries` variables. We will check for cases where multiple different values refer to the same genre and multiple different values refer to the same country.

In [91]:
cleaned_titles['genres'].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

It appears that there are no inconsistencies in the `genres` column of `cleaned_titles`; each value refers to a different genre. However, there may be empty strings representing genres, which are not valid data and should be removed.  
let's check if there are rows in cleaned_titles where genres is an empty string:

In [92]:
cleaned_titles.query('genres == ""')

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

Next, we will follow a similar process for the `production_countries` column. We'll use the `value_counts()` method to obtain the frequency of each value appearing in the `production_countries` column.

In [93]:
cleaned_titles['production_countries'].value_counts()

production_countries
US    5651
IN    1610
GB    1068
JP    1046
FR     720
      ... 
GT       1
CU       1
LK       1
NP       1
FO       1
Name: count, Length: 108, dtype: int64

To view the complete results of `value_counts`, Pandas defaults to displaying only the first and last few values due to the large number of unique values. To show the full output, you can set `display.max_rows` to None, removing the row limit.
However, since we only need to see the complete results temporarily for the current `value_counts` call, we can use `option_context` to change the limit temporarily.

In [94]:
with pd.option_context('display.max_rows', None):
    print(cleaned_titles['production_countries'].value_counts())

production_countries
US         5651
IN         1610
GB         1068
JP         1046
FR          720
KR          637
ES          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
NL          102
HK          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
AT           28
IS           28
LU           27
NZ           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

Based on the output above, countries of production are represented using two-letter country codes, except for one instance of "Lebanon".
The country code for Lebanon is "LB", appearing 39 times, indicating inconsistency in the data. "LB" and "Lebanon" both refer to the same country and should be standardized.

In [95]:
cleaned_titles['production_countries'].replace({'Lebanon':'LB'}, inplace=True)
# Check if "Lebanon" still exists
with pd.option_context('display.max_rows', None):
    print(cleaned_titles['production_countries'].value_counts())

production_countries
US    5651
IN    1610
GB    1068
JP    1046
FR     720
KR     637
ES     637
CA     608
DE     383
CN     295
MX     264
IT     224
BR     221
AU     217
TR     195
PH     192
AR     150
ID     149
BE     148
TW     133
NG     131
PL     126
ZA     103
HK     102
NL     102
CO      94
EG      93
DK      89
TH      87
SE      81
LB      71
NO      68
AE      52
IE      49
SG      47
XX      43
IL      42
RU      41
CL      35
CH      33
PS      32
BG      31
MY      30
AT      28
SA      28
IS      28
LU      27
NZ      27
PE      26
RO      25
QA      24
CZ      22
JO      19
FI      18
HU      18
UY      15
MA      15
PT      14
KW      10
KH      10
PK       9
PR       9
UA       8
MT       8
VN       8
LT       7
IR       7
CD       7
SU       7
TN       7
SN       6
AL       6
KE       6
GH       6
IQ       5
MU       5
CY       5
KN       4
GR       4
IO       4
SY       4
TZ       4
MC       4
GL       3
CM       3
AO       3
BS       3
HR       3
BD       3


Additionally, there may be empty strings representing country codes, which are not valid data. However, since the production countries are not critical for the analysis, observations with empty production countries can be retained.

For `original_credits`, inconsistencies may exist in the `role` column. We will check if there are multiple different values referring to the same type of cast or crew member.

In [101]:
cleaned_credits['role'].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

Based on the output above, the `role` column has only two possible values, `ACTOR` or `DIRECTOR`, and there are no inconsistencies. We can convert this column to a `Category` type, which not only saves memory compared to a string type but also indicates that the values are limited to a specific set.

In [102]:
cleaned_credits['role'] = cleaned_credits['role'].astype('category')
cleaned_credits["role"]

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

#### Handling Invalid or Erroneous Data

Use the `describe()` method of a DataFrame to quickly obtain statistical information about numerical columns.

In [104]:
cleaned_titles.describe()

release_year       runtime      seasons    imdb_score  \
count                          16976  16976.000000  5958.000000  16976.000000   
mean   2015-11-15 06:51:44.618284800     80.899859     2.455018      6.514467   
min              1954-01-01 00:00:00      0.000000     1.000000      1.500000   
25%              2015-01-01 00:00:00     45.000000     1.000000      5.800000   
50%              2018-01-01 00:00:00     90.000000     2.000000      6.600000   
75%              2020-01-01 00:00:00    107.000000     3.000000      7.300000   
max              2022-01-01 00:00:00    225.000000    42.000000      9.600000   
std                              NaN     39.595958     2.868720      1.131246   

         imdb_votes  tmdb_popularity    tmdb_score  
count  1.694500e+04     16844.000000  16517.000000  
mean   3.280916e+04        29.393401      6.846939  
min    5.000000e+00         0.600000      1.000000  
25%    7.800000e+02         4.070000      6.200000  
50%    3.508000e+03        10.195000      6.900000  
75%    1.697600e+04        23.639000      7.500000  
max    2.294231e+06      2274.044000     10.000000  
std    1.141368e+05        93.173092      1.078793

It appears that there are no unrealistic numerical values in `cleaned_titles`.
For `cleaned_credits`, since it does not contain variables that represent numerical meanings, there is no need to use `describe()` to check it.

## Organize the data

In [106]:
cleaned_titles.head()

id        title   type  \
1   tm84618  Taxi Driver  MOVIE   
1   tm84618  Taxi Driver  MOVIE   
2  tm154986  Deliverance  MOVIE   
2  tm154986  Deliverance  MOVIE   
2  tm154986  Deliverance  MOVIE   

                                         description release_year  \
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  Intent on seeing the Cahulawassee River before...   1972-01-01   
2  Intent on seeing the Cahulawassee River before...   1972-01-01   
2  Intent on seeing the Cahulawassee River before...   1972-01-01   

  age_certification  runtime    genres production_countries  seasons  \
1                 R      114     drama                   US      NaN   
1                 R      114     crime                   US      NaN   
2                 R      109     drama                   US      NaN   
2                 R      109    action                   US      NaN   
2                 R      109  thriller                   US      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300

In [107]:
cleaned_credits.head()

person_id       id             name                character   role
0      3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1     14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2      7064  tm84618    Albert Brooks                      Tom  ACTOR
3      3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4     48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

Data Organization is closely tied to the direction of analysis. The goal of this data analysis is to organize the average IMDB ratings of films in different genres such as comedy, action, and sci-fi, thereby uncovering actors who have appeared in high-rated films within each genre.
To obtain both genre and actor data simultaneously, we need to merge `cleaned_titles` and `cleaned_credits` using the `id` column as the key, as both tables share the `id` column representing the film ID.

In [113]:
credits_with_titles = cleaned_titles.merge(cleaned_credits, on='id')
credits_with_titles.head()

id        title   type  \
0  tm84618  Taxi Driver  MOVIE   
1  tm84618  Taxi Driver  MOVIE   
2  tm84618  Taxi Driver  MOVIE   
3  tm84618  Taxi Driver  MOVIE   
4  tm84618  Taxi Driver  MOVIE   

                                         description release_year  \
0  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  A mentally unstable Vietnam War veteran works ...   1976-01-01   
3  A mentally unstable Vietnam War veteran works ...   1976-01-01   
4  A mentally unstable Vietnam War veteran works ...   1976-01-01   

  age_certification  runtime genres production_countries  seasons    imdb_id  \
0                 R      114  drama                   US      NaN  tt0075314   
1                 R      114  drama                   US      NaN  tt0075314   
2                 R      114  drama                   US      NaN  tt0075314   
3                 R      114  drama                   US      NaN  tt0075314   
4                 R      114  drama                   US      NaN  tt0075314   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score person_id  \
0         8.2    808582.0           40.965       8.179      3748   
1         8.2    808582.0           40.965       8.179     14658   
2         8.2    808582.0           40.965       8.179      7064   
3         8.2    808582.0           40.965       8.179      3739   
4         8.2    808582.0           40.965       8.179     48933   

              name                character   role  
0   Robert De Niro            Travis Bickle  ACTOR  
1     Jodie Foster            Iris Steensma  ACTOR  
2    Albert Brooks                      Tom  ACTOR  
3    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  
4  Cybill Shepherd                    Betsy  ACTOR

Once connected, we will have access to specific information about the films and TV shows in which each cast and crew member has participated.

Since we are only interested in exploring the reputation of actors based on their roles in films, and directors are not within our scope of analysis, we will filter the observations where the role is 'ACTOR' for further analysis.

In [114]:
actor_with_titles = credits_with_titles.query('role == "ACTOR"')

To uncover actors in high-rated IMDB films within each genre, we need to group the data by genre and actor.
When grouping actors, we use the `person_id` instead of the `name` variable. This is because names can be misspelled or duplicated, whereas the cast and crew ID more accurately reflects the specific actor.

In [116]:
groupby_genres_and_person_id = actor_with_titles.groupby(['genres', 'person_id'])

After grouping, we only need to perform aggregation on the `imdb_score` values. Therefore, we will extract the `imdb_score` variable and then call `mean` to calculate the average IMDB rating of the films each actor has participated in within each genre.

In [125]:
imdb_score_groupby_genres_and_person_id = groupby_genres_and_person_id['imdb_score'].agg('mean')
imdb_score_groupby_genres_and_person_id

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

We can call `reset_index` to flatten the hierarchical index and obtain a more organized DataFrame.

In [126]:
imdb_score_groupby_genres_and_person_id_df = imdb_score_groupby_genres_and_person_id.reset_index()
imdb_score_groupby_genres_and_person_id_df

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

Now that the IMDB ratings data grouped by genre and actor is organized, we can proceed to the next steps of the analysis.
We can further refine the data by grouping the results again to find the highest average rating for actors' works in each genre and the corresponding actor's name.
To achieve this, we need to group by `genres` again, then extract the `imdb_score` variable and compute its maximum value.

In [127]:
genres_max_scores = imdb_score_groupby_genres_and_person_id_df.groupby("genres")["imdb_score"].max()
genres_max_scores

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

Once we know the highest score, we can join this result with the previously obtained `imdb_score_groupby_genres_and_person_id_df` to find out which actor IDs correspond to the highest scores. In other words, we will determine which actor achieved the highest average rating.

In [131]:
genres_max_score_with_person_id = pd.merge(genres_max_scores,imdb_score_groupby_genres_and_person_id_df, on=['genres','imdb_score'])
genres_max_score_with_person_id

genres  imdb_score person_id
0     action         9.3     12790
1     action         9.3      1303
2     action         9.3     21033
3     action         9.3    336830
4     action         9.3     86591
..       ...         ...       ...
131      war         8.8    826547
132  western         8.9     22311
133  western         8.9     28166
134  western         8.9     28180
135  western         8.9     93017

[136 rows x 3 columns]

We can see that the highest score might correspond to multiple actors, as several actors may have the same average rating.
To get the actor names corresponding to the actor IDs, we can join this with the `cleaned_credits` DataFrame. Since this DataFrame has other columns, we only need the `person_id` and `name` columns. We can extract these two columns and remove duplicate rows.

In [132]:
actor_id_with_names = cleaned_credits[['person_id', 'name']].drop_duplicates()
actor_id_with_names.head(10)

person_id             name
0      3748   Robert De Niro
1     14658     Jodie Foster
2      7064    Albert Brooks
3      3739    Harvey Keitel
4     48933  Cybill Shepherd
5     32267      Peter Boyle
6    519612   Leonard Harris
7     29068   Diahnne Abbott
8    519613      Gino Ardito
9      3308  Martin Scorsese

The next step is to join `actor_id_with_names` with the previously obtained `genres_max_score_with_person_id`, adding the `name` variable to display the names of the actors with the highest average ratings.

In [134]:
genres_max_score_with_actor_name = genres_max_score_with_person_id.merge(actor_id_with_names, on='person_id')
genres_max_score_with_actor_name

genres  imdb_score person_id                name
0       action         9.3     12790         Olivia Hack
1        scifi         9.3     12790         Olivia Hack
2       action         9.3      1303       Jessie Flower
3    animation         9.3      1303       Jessie Flower
4       family         9.3      1303       Jessie Flower
..         ...         ...       ...                 ...
131        war         8.8    826547         Yuto Uemura
132    western         8.9     22311     Koichi Yamadera
133    western         8.9     28166  Megumi Hayashibara
134    western         8.9     28180      Unsho Ishizuka
135    western         8.9     93017            Aoi Tada

[136 rows x 4 columns]

To group the same genres together, we can use the `sort_values` method to sort the rows in the result by `genres`. Then, we can use `reset_index` to reorder the indices.
After resetting the indices, the DataFrame will have an additional `index` column, which we can delete.

In [140]:
genres_max_score_with_actor_name = genres_max_score_with_actor_name.sort_values("genres").reset_index().drop("index", axis=1)
genres_max_score_with_actor_name

genres  imdb_score person_id                name
0     action         9.3     12790         Olivia Hack
1     action         9.3    336830     André Sogliuzzo
2     action         9.3     21033          Zach Tyler
3     action         9.3     86591       Cricket Leigh
4     action         9.3      1303       Jessie Flower
..       ...         ...       ...                 ...
131      war         8.8    826547         Yuto Uemura
132  western         8.9     28180      Unsho Ishizuka
133  western         8.9     22311     Koichi Yamadera
134  western         8.9     28166  Megumi Hayashibara
135  western         8.9     93017            Aoi Tada

[136 rows x 4 columns]